### PR TITLE
Feature/lol/itsi saved blank project

### PR DESCRIPTION
**Pull request by lliss**
In some cases we want a project to be saved as soon as the user lands on the
site. This is determined by the addition of a URL query parameter. If this
parameter is set, a project will be instantly created and saved. If the user
designates an area of interest and then models it everything will be saved. If
they then move back out and redraw, the same project will be used and the data
will be overwritten.

Connects #561 

To test:

 * Alter `src/mmw/apps/home/views.py` and set `EMBED_FLAG: True` in `def get_client_settings(request):`
 * Visit localhost:8000
 * Check DB, you should see the new project has been save.
 * Draw AOI, and move to modeling page.
 * Create scenarios as normal and ensure they are saved.
 * Back up to analyze mode and return to the project. Everything should remain unchanged.
 * Back up the initial screen and then navigate back to the project. Everything should remain unchanged. 
 * Back up to the main drawing page and click a drawing tool.
 * You should be prompted about the possible hazards.
 * Cancel out. Move back to project page. The project should be unchanged.
 * Back up to the main drawing page and click a drawing tool.
 * Agree to make the changes.
 * Move back to modeling page. The new data should be saved.
 * Create scenarios as normal and ensure they are saved.
 * Reload page at the project URL and ensure that it properly reloads.
 * After reloading, you should remain in "activity mode" which has the same properties as described above.
 * Disable the hard coded value in `src/mmw/apps/home/views.py` 
 * Reload the URL for the project you created. This should have an is_activity flag set on it. Loading this should load the page in "activity mode."

### CODE REVIEW

File: src/mmw/js/src/core/utils.js [hunk 1]
```diff
@@ -31,6 +31,17 @@ var utils = {
         return params;
     },
 
+    /**
+     * Returns the value of a query param from the URL if it exists or null:
+     * https://stackoverflow.com/questions/901115
+     */
+    getParameterByName: function(name) {
```
Comment (rajadain) [inline src/mmw/js/src/core/utils.js]: I like this function a lot better than the `parseQueryString` above. We might consider removing `parseQueryString` entirely and replacing its one usage here: https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/js/src/user/controllers.js#L23 with something like this:

```javascript
var nextParam = utils.getParameterByName('next');
...
model = new models.ItsiSignUpFormModel({
    ...
    next: nextParam ? nextParam : '/'
});
```

File: src/mmw/js/src/draw/controllers.js [hunk 2]
```diff
@@ -27,6 +29,8 @@ var DrawController = {
         App.rootView.geocodeSearchRegion.show(geocodeSearch);
         App.rootView.drawToolsRegion.show(toolbarView);
 
+        enableSingleProjectMode();
```
Comment (rajadain) [inline src/mmw/js/src/draw/controllers.js]: This function's name seems to imply that Single Project Mode is _always_ enabled. Consider renaming to `enableSingleProjectModeIfItsi`, or to place the condition here instead of within the function:

```javascript
if (coreUtils.getParameterByName('initialize_itsi') === 'true') {
    enableSingleProjectMode();
}
```

File: src/mmw/js/src/draw/controllers.js [hunk 3]
```diff
@@ -51,6 +55,28 @@ var DrawController = {
...
 };
 
+/**
+ * If itsi flag is set, prepare a project immedialty upon visiting the page.
+ * This will be the only project the user can save during this session.
+ */
+function enableSingleProjectMode() {
+    if (coreUtils.getParameterByName('initialize_itsi') === 'true') {
+
+        App.singleProjectMode = true;
```
Comment (rajadain) [inline src/mmw/js/src/draw/controllers.js]: Consider explicitly setting this to `false` by default.

File: src/mmw/js/src/modeling/controllers.js [hunk 4]
```diff
@@ -43,40 +44,65 @@ var ModelingController = {
...
+
+                // Clear current scenarios and start over.
+                // Must convert to an array first to avoid conflicts with
+                // collection events that are disassociating the model during
+                // the loop.
+                var locks = [];
+                _.each(project.get('scenarios').toArray(), function(model) {
+                    var $lock = $.Deferred();
+                    locks.push($lock);
+                    model.destroy({
```
Comment (rajadain) [inline src/mmw/js/src/modeling/controllers.js]: Since having once gone back from the `/project` view to `/analyze`, even if I make no changes, all the Scenarios are killed, I'm afraid that people will easily make mistakes and end up deleting a lot of work. I think we should not delete scenarios unless the AoI has changed, so if one goes from `/project` to `/analyze` to `/project` without making any changes, their old work is preserved.

Also, when resetting the AoI, the user should be shown a Confirm Modal that informs them of scenario deletion, and gives them a chance to cancel that.

**Comment by rajadain** [comment]
Great work. I think we can follow up with the changes talked about with @mmcfarland. Have a look at [`middleware.py`](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/apps/user/middleware.py#L33) and `client_settings` in [`views.py`](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/apps/home/views.py) for ideas about stringing through a flag from Django to Backbone without query parameter parsing.

File: src/mmw/js/src/core/views.js [hunk 5]
```diff
@@ -414,6 +459,135 @@ var ModificationPopupView = Marionette.ItemView.extend({
...
+        this.ui.input.focus().select();
+    },
+
+    signIn: function() {
+        this.options.app.getUserOrShowLogin();
+    }
+
+});
+
+>>>>>>> Trigger event when canceling a confirmation modal.
```
Comment (rajadain) [inline src/mmw/js/src/core/views.js]: Can't really test this until merge conflicts are resolved.

**Comment by rajadain** [comment]
Having a look now.

**Comment by rajadain** [comment]
+1, functionally works very well. Great work! By the way, we don't have to hard-set the value of `EMBED_FLAG` in `home/views.py`, we can just visit the URL with `/?itsi_embed=true` which will enable it (and be a more accurate test).

Tested that project is saved as soon as one lands on the draw page when `activityMode` is enabled, that one can go back and forth at ease without harming the Scenarios, that one is prompted a warning when resetting the Area of Interest, that "Cancel" and "Make Changes" in that warning work as expected, and that loading a project that was started as an Activity turns `activityMode` on.

A loophole I found is when a project that is marked as an activity is made public, and accessed by a different / anonymous user. Since `activityMode` is still set to `true`, they can go back and change the AoI, which gives them the same warning, but when they try to proceed again the `DELETE` scenario requests are rejected with `403`s. I think that `activityMode` should only be set if the current user is the project owner, but perhaps we need to think more about this.

Going to look through the code now.

File: src/mmw/js/src/core/views.js [hunk 6]
```diff
@@ -1,6 +1,7 @@
 "use strict";
 
-var $ = require('jquery'),
+var Backbone = require('../../shim/backbone'),
```
Comment (rajadain) [inline src/mmw/js/src/core/views.js]: `Backbone` is not used in this file, is causing lint errors.

**Comment by lliss** [comment]
@azavea-bot rebuild

File: src/mmw/js/src/core/views.js (hunk 6 above)
Comment (lliss) [inline src/mmw/js/src/core/views.js]: Fixed and rebased.

**Comment by lliss** [comment]
@azavea-bot rebuild

File: src/mmw/js/src/core/views.js [hunk 7]
```diff
@@ -203,6 +210,45 @@ var MapView = Marionette.ItemView.extend({
         }
     },
 
+    aoiChangeWarning: _.debounce(function() {
```
Comment (rajadain) [inline src/mmw/js/src/core/views.js]: Cool use of `_.debounce` here, I didn't know we could define view methods this way.

File: src/mmw/js/src/app.js [hunk 8]
```diff
@@ -13,11 +14,19 @@ var App = new Marionette.Application({
         this.restApi = new RestAPI();
         this.map = new models.MapModel();
 
+        // If in embed mode we are be default in activity mode.
+        var activityMode = settings.get('itsi_embed');
+        settings.set('activityMode', activityMode);
```
Comment (rajadain) [inline src/mmw/js/src/app.js]: Consider renaming to `activity_mode` to better match other setting keys.

File: src/mmw/js/src/draw/controllers.js [hunk 9]
```diff
@@ -51,6 +55,29 @@ var DrawController = {
...
+ * If we are in embed mode then the project is an activity and we want to keep
+ * the same project reguardless of changes to the AOI. This prepares a project
+ * immedialty upon visiting the page and will be the only project the user can
+ * save during this session.
+ */
+function enableSingleProjectModeIfActivity() {
+    if (settings.get('activityMode')) {
+        if (!App.currProject) {
+            var project = new modelingModels.ProjectModel({
+                name: 'New Project',
```
Comment (rajadain) [inline src/mmw/js/src/draw/controllers.js]: New projects are named "Untitled Project" by default in the regular case.

File: src/mmw/js/src/core/settings.js [hunk 10]
```diff
@@ -5,15 +5,24 @@ var defaultSettings = {
...
+        return settings[key];
+    } catch (exc) {
+        return undefined;
+    }
 }
 
 module.exports = {
-    getSettings: getSettings,
-    setSettings: setSettings
+    get: get,
```
Comment (rajadain) [inline src/mmw/js/src/core/settings.js]: Thanks, I like this refactor a lot better than the original implementation.

File: src/mmw/js/src/core/modals/views.js [hunk 11]
```diff
@@ -55,16 +55,22 @@ var ConfirmView = ModalBaseView.extend({
...
+        'click @ui.confirmation': 'primaryAction',
+        'click @ui.deny': 'cancelAction'
     }, ModalBaseView.prototype.events),
 
     primaryAction: function() {
         this.triggerMethod('confirmation');
         this.hide();
+    },
+
+    cancelAction: function() {
```
Comment (rajadain) [inline src/mmw/js/src/core/modals/views.js]: A similar method exists in `user/views.js` where it is called `dismissAction`.

This function only works when the user clicks cancel. If they click outside the modal, or use <kbd>Escape</kbd>, then the app is in an inconsistent state where neither event was triggered:

mmw-clear-aoi-issue

Consider refactoring so that instead of listening to the `deny` event, the consumer listens for the `hidden.bs.modal` and checks for a flag that indicates if `primaryAction` was executed or not.

File: src/mmw/js/src/app.js (hunk 8 above)
Comment (lliss) [inline src/mmw/js/src/app.js]: Trying to stick with camelCase throughout the application where possible we use snake_case where we are getting attributes directly from the API so we don't have to do nasty conversion of key formats. I think that was decided to be the lesser evil though I think that is the exception rather than the rule.

File: src/mmw/js/src/draw/controllers.js (hunk 9 above)
Comment (lliss) [inline src/mmw/js/src/draw/controllers.js]: Originally I had this named in a way that would differentiate it from the normal projects. This is generic enough that I should either make it more specific such as "New Activity" or just make it the same "Untitled Project" we are using elsewhere.

File: src/mmw/js/src/core/modals/views.js (hunk 11 above)
Comment (lliss) [inline src/mmw/js/src/core/modals/views.js]: Oh that's an annoying issue. I might make a separate card for that.

File: src/mmw/js/src/app.js (hunk 8 above)
Comment (rajadain) [inline src/mmw/js/src/app.js]: Cool, makes sense.

File: src/mmw/js/src/draw/controllers.js (hunk 9 above)
Comment (rajadain) [inline src/mmw/js/src/draw/controllers.js]: Since we have the `is_activity` flag, it allows us (the backend people) to easily differentiate activities and projects, and it isn't different enough to enable users to clearly differentiate between the two. I like the "New Activity" idea, which is clear and explicit.

Any thoughts on this, @mmcfarland ?

Comment (rajadain) [inline src/mmw/js/src/draw/controllers.js]: Thanks for clarifying the thought process though, it makes sense and I appreciate it. I made my original comment to ensure this wasn't an oversight.

File: src/mmw/js/src/core/modals/views.js (hunk 11 above)
Comment (lliss) [inline src/mmw/js/src/core/modals/views.js]: Fixed in efb8daf

**Comment by lliss** [comment]
Fixed the issue with the failing DELETE requests in 4cf078e.

**Comment by rajadain** [comment]
Taking another look

File: src/mmw/js/src/core/modals/views.js [hunk 12]
```diff
@@ -14,8 +14,18 @@ var _ = require('underscore'),
...
+            feedbackRequired = {
+                'data-backdrop': 'static',
+                'data-keyboard': 'false'
+            };
+
+        // If feedback is required, we don't allow the user to get away from
+        // the modal without clicking confirm or deny. This way we ensure we
+        // get an event one way or another.
+        // http://getbootstrap.com/javascript/#modals-options
+        return this.model.get('feedbackRequired') ? _.extend(defaults, feedbackRequired) : defaults;
```
Comment (rajadain) [inline src/mmw/js/src/core/modals/views.js]: Very cool, I love this solution!

File: src/mmw/js/src/modeling/controllers.js [hunk 13]
```diff
@@ -41,42 +43,82 @@ var ModelingController = {
                         project.get('scenarios').makeFirstScenarioActive();
                     }
                     project.getResultsIfNeeded();
+
+                    // If this project is an activity then the application's
+                    // behaior changes but only if the project is owned by the
```
Comment (rajadain) [inline src/mmw/js/src/modeling/controllers.js]: Typo, should be "behavior" instead of "behaior"

**Comment by rajadain** [comment]
If going back to existing AoI and selecting "Boundary Layers", they are still clickable if they are within the AoI, and lead to an inconsistent state:

mmw-clear-aoi-issue-3

But the good thing is that the inconsistency isn't persisted, as seen in the above GIF. The only thing it might affect are the model results for which we send the AoI each time, and they aren't recalculated when loading the project next time because the inputs and modifications are the same, but the AoI is different.

If this is not easily resolved, we might make a separate card for this, since it doesn't seem to be too damaging / not too easily reproduced.

File: src/mmw/js/src/modeling/controllers.js [hunk 14]
```diff
@@ -41,42 +43,82 @@ var ModelingController = {
...
+                    // current user.
+                    if (App.user.loggedInUserMatch(project.get('user_id')) && project.get('is_activity')) {
```
Comment (rajadain) [inline src/mmw/js/src/modeling/controllers.js]: We cannot use `App.user.loggedInUserMatch` here because the `App.user` isn't initialized until later. In the following example, I revisit my own project made using `itsi_embed=true` in non-embedded mode, and even though it is an activity and I own it, `activityMode` is `false`:

image

**Comment by lliss** [comment]
Okay. I removed the test to ensure the user owns the activity. That's an edge case and we'll address it in a new issue. Same for the boundary layer selector not being properly discarded.

**Comment by rajadain** [comment]
Alright, going in for one final pass

**Comment by rajadain** [comment]
Very minor style recommendation: consider constraining commit body to 79 characters to prevent odd wrapping in GitHub:

image

File: src/mmw/js/src/app.js [hunk 15]
```diff
@@ -13,11 +14,19 @@ var App = new Marionette.Application({
         this.restApi = new RestAPI();
         this.map = new models.MapModel();
 
+        // If in embed mode we are be default in activity mode.
```
Comment (rajadain) [inline src/mmw/js/src/app.js]: Typo, "be default" should be "by default"

File: src/mmw/js/src/core/models.js [hunk 16]
```diff
@@ -46,6 +47,18 @@ var MapModel = Backbone.Model.extend({
         }
     },
 
+    stashAOI: function() {
+        // Since we oscilate between an area of interest and a blank map, stash
```
Comment (rajadain) [inline src/mmw/js/src/core/models.js]: Typo, "oscilate" should be "oscillate"

File: src/mmw/js/src/core/views.js [hunk 17]
```diff
@@ -10,7 +10,9 @@ var $ = require('jquery'),
     modificationConfigUtils = require('../modeling/modificationConfigUtils'),
     headerTmpl = require('./templates/header.html'),
     modificationPopupTmpl = require('./templates/modificationPopup.html'),
-    areaOfInterestTmpl = require('../core/templates/areaOfInterestHeader.html'),
```
Comment (rajadain) [inline src/mmw/js/src/core/views.js]: Hah this was a funny self-reference. Great catch fixing that.

**Comment by lliss** [comment]
I always thought it wrapped at 80.

**Comment by rajadain** [comment]
+1 great work, great code. Works very well. I hope we have to do minimal modifications to this as we go along, although I suspect that as we start embedding it inside ITSI we may come across some edge cases that we didn't think about here. But this is good to merge (once the conflicts have been resolved).

I did find one another issue regarding changing the AoI, making a separate card for that. Will reference it when ready.

**Comment by rajadain** [comment]
Added #638 for the above mentioned issue.
